### PR TITLE
Refactored the grab pass and scene depth rendering

### PR DESCRIPTION
**Pull request by mvaligursky**
Fixes: https://github.com/playcanvas/engine/issues/4090
Related: https://github.com/playcanvas/engine/issues/2953

This PR refactors and improved both grab pass functionality, as well as the scene depth rendering. Files grab-pass.js and scene-depth.js were refactored into scene-grab.js. The main motivation is to minimize number of resolves and buffer copies taking place, and keeping to more formal render pass implementation required for webGPU.

**Shared changes**
- previously, if the depth texture was used, a depth layer was enabled, and all cameras that had the depth layer would capture the depth, wasting performance for cameras other than the main one. This is now controlled by camera (details below)
- there are now new and consistent textures names for these two textures. Note that even though the engine uses new names, old names are still supported:
```
uDepthMap -> uSceneDepthMap
texture_grabPass -> uSceneColorMap
```

**Changes to grab-pass**
- previously, when a shader referenced grab pass texture, a rendering would be interrupted in the middle of draw call, a framebuffer copied to texture and then the texture used for the draw call. This gets very expensive when multiple materials use the texture. The commonly used solution is to do a single framebuffer copy after opaque meshes are rendered, making it available for the transparent meshes / post-processing. In our case, this executed during the depth layer, giving used the control over the time inside the frame.
- There is new public API to request a camera to capture the scene: 
```
CameraComponent.requestSceneColorMap
```

**Changes to depth pass**
- There is new public API to request a camera to capture the scene. This is automatically requested by post-effects, but needs to be manually requested for other uses (for example Soft-Particles).
```
CameraComponent.requestSceneDepthMap
```

### CODE REVIEW

File: src/framework/components/camera/component.js [hunk 1]
```diff
@@ -144,6 +146,20 @@ class CameraComponent extends Component {
      */
     onPostRender = null;
 
+    /**
+     * A counter of requests of depth map rendering.
+     *
+     * @type {number}
```
Comment (willeastcott) [inline src/framework/components/camera/component.js:152]: Does this need a @private?

File: src/framework/components/camera/component.js [hunk 2]
```diff
@@ -144,6 +146,20 @@ class CameraComponent extends Component {
...
+     */
+    _renderSceneDepthMap = 0;
+
+    /**
+     * A counter of requests of color map rendering.
+     *
+     * @type {number}
```
Comment (willeastcott) [inline src/framework/components/camera/component.js:159]: Does this need a @private tag?

File: src/framework/components/camera/component.js (hunk 1 above)
Comment (mvaligursky) [reply src/framework/components/camera/component.js:152]: well spotted, thanks, I keep forgetting anything documented is public by default now. Done.

File: src/framework/components/camera/component.js (hunk 2 above)
Comment (mvaligursky) [reply src/framework/components/camera/component.js:159]: and done.

File: src/framework/scene-grab.js [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+// Internal class abstracting the access to the depth and color texture of the scene.
+// color frame buffer is copied to a texture
+// For webgl 2 devices, the depth buffer is copied to a texture
+// for webgl 1 devices, the scene's depth is rendered to a separate RGBA texture
+//
+// TODO: implement mipmapped color buffer support for WebGL 1 as well, which requires
+// the texture to be a power of two, by first downscaling the captured framebuffer
+// texture to smaller power of 2 texture, and then generate mipmaps and use it for rendering
+// TODO: or even better, implement blur filter to have smoother lower levels
```
Comment (willeastcott) [inline src/framework/scene-grab.js:30]: Any reason not to make this a JSDoc block with an `@ignore` tag?

File: src/framework/scene-grab.js [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+            this.initWebGl1();
+        }
+    }
+
+    allocateTexture(device, name, format, isDepth, mipmaps) {
+
+        // allocate texture that will store the depth
+        const texture = new Texture(device, {
+            name,
+            format: format,
```
Comment (willeastcott) [inline src/framework/scene-grab.js:59]: Interesting - you're starting to use object shorthand....but I would use it on all properties or none. So maybe update this one to just `format`?

File: src/framework/scene-grab.js (hunk 3 above)
Comment (mvaligursky) [reply src/framework/scene-grab.js]: Good point, done.

File: src/framework/scene-grab.js (hunk 4 above)
Comment (mvaligursky) [reply src/framework/scene-grab.js]: and done as well, I'm warming up to it ;)

**Review by willeastcott — APPROVED** [review]

**Comment by NewboO** [comment]
Hello,

How to get the old behaviour for the color buffer? In some of my projects, I'm using shaders to blend UI sprites with custom functions (similar to photoshop blend modes) so I need the current color buffer when rendering the sprite, not one captured some time before because as those sprites can overlap each other they're now cancelling each other instead of stacking.

**Comment by mvaligursky** [comment]
@NewboO - Unfortunately we don't have a support for this anymore. For multiple reason, this has been removed for now. We will consider having this as an option in the future, but most likely not very immediately.

**Comment by yaustar** [comment]
@NewboO, would you be able to share an example shader that no longer works? Perhaps we can provide an alternative path?

**Comment by NewboO** [comment]
Here's a sample project: https://playcanvas.com/project/950042/overview/color-buffer

And here's how it looks with previous engine version: https://launch.playcanvas.com/1456173?debug=true&version=1.53.4
image

Now with current engine it gives this: https://launch.playcanvas.com/1456173?debug=true
image

This is a very basic example with only a difference blending on a single layer but I'd like to be able to do this kind of thing with more complex blending formulas on any layer.

**Comment by mvaligursky** [comment]
I think a possible way to do something like this would be to create multiple render targets .. and swap rendering between them, so that you can use what was already rendered to as a texture for following rendering. Not as convenient, especially if this is in 3D with dynamically changing orders unfortunately, so more of a limited case solution I'd say.

**Comment by NewboO** [comment]
An update if anyone else than me is interested in getting the old behaviour: the best workaround I found was to duplicate the SceneGrab code I need and call it in a custom Layer.onDrawCall function. It's not perfect but good enough for my use cases and it doesn't require juggling with render targets.

My sample project is updated with a dirty PoC showing both behaviours.

**Pull request merged**